# Result È™i ?
`Result<T>` este o enumerare folositÄƒ pentru a descrie posibilele erori. Are douÄƒ variante:

- Ok(T): A fost gÄƒsitÄƒ o valoare T
- Err(e): A fost gÄƒsitÄƒ o eroare cu valoarea `e`
Pe scurt, rezultatul aÈ™teptat este `Ok`, Ã®n timp ce rezultatul neaÈ™teptat este `Err`.

1. ğŸŒŸğŸŒŸ
```rust,editable

// UMPLE blank-urile È™i REZOLVÄ‚ erorile
use std::num::ParseIntError;

fn multiply(n1_str: &str, n2_str: &str) -> __ {
    let n1 = n1_str.parse::<i32>();
    let n2 = n2_str.parse::<i32>();
    Ok(n1.unwrap() * n2.unwrap())
}

fn main() {
    let result = multiply("10", "2");
    assert_eq!(result, __);

    let result = multiply("t", "2");
    assert_eq!(result.__, 8);

    println!("Success!");
}
```

### ? 
`?` este aproape echivalent cu `unwrap`, dar `?` Ã®ntoarce rezultatul Ã®n loc sÄƒ provoace panicÄƒ la `Err`.

2. ğŸŒŸğŸŒŸ
```rust,editable

use std::num::ParseIntError;

// IMPLEMENTEAZÄ‚ multiply cu ?
// NU folosi unwrap aici
fn multiply(n1_str: &str, n2_str: &str) -> __ {
}

fn main() {
    assert_eq!(multiply("3", "4").unwrap(), 12);
    println!("Success!");
}
```

3. ğŸŒŸğŸŒŸ
```rust,editable

use std::fs::File;
use std::io::{self, Read};

fn read_file1() -> Result<String, io::Error> {
    let f = File::open("hello.txt");
    let mut f = match f {
        Ok(file) => file,
        Err(e) => return Err(e),
    };

    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

// UMPLE blank-urile cu o linie de cod
// NU schimba nicio linie de cod
fn read_file2() -> Result<String, io::Error> {
    let mut s = String::new();

    __;

    Ok(s)
}

fn main() {
    assert_eq!(read_file1().unwrap_err().to_string(), read_file2().unwrap_err().to_string());
    println!("Success!");
}
```

### map & and_then
[map](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.map) È™i [and_then](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.and_then) sunt doi combinatori comuni pentru `Result<T, E>` (È™i pentru `Option<T>`).

4. ğŸŒŸğŸŒŸ 

```rust,editable
use std::num::ParseIntError;

// UMPLE blank-ul Ã®n douÄƒ moduri: map È™i and_then
fn add_two(n_str: &str) -> Result<i32, ParseIntError> {
   n_str.parse::<i32>().__
}

fn main() {
    assert_eq!(add_two("4").unwrap(), 6);

    println!("Success!");
}
```

5. ğŸŒŸğŸŒŸğŸŒŸ
```rust,editable
use std::num::ParseIntError;

// Cu tipul returnat rescris, folosim potrivirea modelelor fÄƒrÄƒ `unwrap()`.
// Dar este atÃ¢t de detaliat...
fn multiply(n1_str: &str, n2_str: &str) -> Result<i32, ParseIntError> {
    match n1_str.parse::<i32>() {
        Ok(n1)  => {
            match n2_str.parse::<i32>() {
                Ok(n2)  => {
                    Ok(n1 * n2)
                },
                Err(e) => Err(e),
            }
        },
        Err(e) => Err(e),
    }
}

// Rescrierea `multiply` pentru a o face concisÄƒ
// Ar trebui sÄƒ foloseÈ™ti ATÃ‚T `and_then` cÃ¢t È™i `map` aici.
fn multiply1(n1_str: &str, n2_str: &str) -> Result<i32, ParseIntError> {
    // IMPLEMENTEAZÄ‚...
}

fn print(result: Result<i32, ParseIntError>) {
    match result {
        Ok(n)  => println!("n is {}", n),
        Err(e) => println!("Error: {}", e),
    }
}

fn main() {
    // Aceasta oferÄƒ Ã®ncÄƒ un rÄƒspuns rezonabil.
    let twenty = multiply1("10", "2");
    print(twenty);

    // UrmÄƒtorul oferÄƒ acum un mesaj de eroare mult mai util.
    let tt = multiply("t", "2");
    print(tt);

    println!("Success!");
}
```

### Alias de tipuri
Folosirea `std::result::Result<T, ParseIntError>` peste tot este stufos È™i plictisitor, putem folosi un alias Ã®n acest scop.

La nivel de modul, crearea de aliasuri poate fi deosebit de utilÄƒ. Erorile gÄƒsite Ã®ntr-un modul specific au adesea acelaÈ™i tip `Err`, astfel cÄƒ un singur alias poate defini concis toate `Result`-urile asociate. Aceasta este atÃ¢t de utilÄƒ Ã®ncÃ¢t biblioteca standard furnizeazÄƒ una: [`io::Result`](https://doc.rust-lang.org/std/io/type.Result.html).

6. ğŸŒŸ
```rust,editable
use std::num::ParseIntError;

// UMPLE blank-ul
type __;

// FoloseÈ™te aliasul de mai sus pentru a te referi la tipul nostru specific `Result`.
fn multiply(first_number_str: &str, second_number_str: &str) -> Res<i32> {
    first_number_str.parse::<i32>().and_then(|first_number| {
        second_number_str.parse::<i32>().map(|second_number| first_number * second_number)
    })
}

// Aici, aliasul permite din nou sÄƒ economisim spaÈ›iu.
fn print(result: Res<i32>) {
    match result {
        Ok(n)  => println!("n is {}", n),
        Err(e) => println!("Error: {}", e),
    }
}

fn main() {
    print(multiply("10", "2"));
    print(multiply("t", "2"));

    println!("Success!");
}
```

### Utilizarea Result Ã®n `fn main`
De obicei, funcÈ›ia `main` va arÄƒta aÈ™a:
```rust
fn main() {
    println!("Hello World!");
}
```

Cu toate acestea, main poate avea È™i un tip de returnare `Result`. DacÄƒ apare o eroare Ã®n cadrul funcÈ›iei `main`, aceasta va returna un cod de eroare È™i va imprima o reprezentare de depanare a erorii (traitul Debug).

UrmÄƒtorul exemplu aratÄƒ o astfel de situaÈ›ie:
```rust,editable

use std::num::ParseIntError;

fn main() -> Result<(), ParseIntError> {
    let number_str = "10";
    let number = match number_str.parse::<i32>() {
        Ok(number)  => number,
        Err(e) => return Err(e),
    };
    println!("{}", number);
    Ok(())
}
```

> PuteÈ›i gÄƒsi soluÈ›iile [aici](https://github.com/sunface/rust-by-practice) (Ã®n cadrul cÄƒii soluÈ›iilor), dar folosiÈ›i-le doar atunci cÃ¢nd aveÈ›i nevoie. :)
